File: ImasLiveDB-Android/app/src/main/kotlin/com/fugaif/imaslivedb/data/games/GameProgressStore.kt
package com.fugaif.imaslivedb.data.games

import android.content.Context
import com.fugaif.imaslivedb.data.model.DailyPick
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject
import uniffi.imas_core.GameProgressUpdate
import uniffi.imas_core.GameRecord
import uniffi.imas_core.GameStreakState
import uniffi.imas_core.gameProgressApplyResult
import uniffi.imas_core.gameProgressDidClearToday
import uniffi.imas_core.gameProgressDisplayStreak

/** ハブが束ねるゲームの識別子。name は永続キー兼用なので変更しない。iOS GameKind の移植。 */
enum class GameKind(val displayName: String, val scoreIsPercent: Boolean) {
    introDon("イントロドン", false),
    idolQuiz("アイドル当てクイズ", false),
    songSingerQuiz("ソロ曲クイズ", false),
    colorMatch("カラーマッチ", true)
}

/**
 * 未プレイの初期値。保存値の型は imas-core の [GameRecord] そのもの
 * (uniffi 生成なので Kotlin の既定引数を持てず、初期値をここで名付ける)。
 */
fun emptyGameRecord(): GameRecord =
    GameRecord(lastScore = 0, lastOutOf = 0, bestScore = 0, bestOutOf = 0, playCount = 0)

/** 未達成の初期値。理由は [emptyGameRecord] と同じ。 */
private fun emptyStreakState(): GameStreakState =
    GameStreakState(streak = 0, totalDays = 0, lastClearedDay = null)

/**
 * 1 度でも遊んだか。コアの `GameRecord::has_played` と同じ規則だが、
 * 述語 1 行のために FFI 面を増やさず Kotlin 側の拡張として持つ。
 */
val GameRecord.hasPlayed: Boolean get() = playCount > 0

/**
 * ゲーム横断のローカル進捗ストア (SharedPreferences)。iOS GameProgressStore の移植。
 *
 * 保存の実体 (キー `game_records_v1` / `game_streak_v1` の JSON) だけがここの責務で、
 * **更新規則は imas-core の `domain::game_progress`** が持つ。ストアは
 * 「読む → コアに渡す → 返ってきた値を書く」に痩せている。
 *
 * 日付キーの生成も [DailyPick] 経由でコアに委ねる (書式を自前で組むと iOS と
 * 食い違い、保存済みの連続記録キーと突き合わなくなる)。連続達成の単位は
 * 「そのユーザーの 1 日」なので、公演日の比較に使う JST 固定の
 * [com.fugaif.imaslivedb.data.model.JstDay] とは意味が違い、統合してはいけない。
 */
class GameProgressStore(context: Context) {

    private val prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private val _records = MutableStateFlow(loadRecords())
    val records: StateFlow<Map<GameKind, GameRecord>> = _records.asStateFlow()

    private val _streak = MutableStateFlow(loadStreak())
    val streak: StateFlow<GameStreakState> = _streak.asStateFlow()

    fun record(kind: GameKind): GameRecord = _records.value[kind] ?: emptyGameRecord()

    /** ストリークが「今日途切れていないか」。表示用 (今日/昨日までクリアなら継続扱い)。 */
    val displayStreak: Int
        get() = gameProgressDisplayStreak(_streak.value, DailyPick.dayKey(), DailyPick.previousDayKey())

    /**
     * 今日デイリーチャレンジを達成済みか。
     *
     * `lastClearedDay == 今日` の比較をストアの外で書くとコアと二重実装になるので、
     * 判定はコアに置いたままこの述語で公開する (iOS `GameProgressStore.didClearToday` と対)。
     */
    val didClearToday: Boolean
        get() = gameProgressDidClearToday(_streak.value, DailyPick.dayKey())

    /**
     * ゲーム結果を記録する。score/outOf は「獲得点 / 満点」。
     *
     * 「自己ベストを更新したか」は best を上書きする**前**に判定する必要があり、その順序ごと
     * コアの 1 呼び出しに畳んである。結果画面のバッジは戻り値の `isNewBest` を使うこと
     * (画面側で判定し直すと順序を崩した瞬間に恒久的に出なくなる)。
     */
    fun recordResult(kind: GameKind, score: Int, outOf: Int): GameProgressUpdate {
        val update = gameProgressApplyResult(
            record = record(kind),
            streak = _streak.value,
            score = score,
            outOf = outOf,
            todayKey = DailyPick.dayKey(),
            yesterdayKey = DailyPick.previousDayKey()
        )
        // 記録として成立しない回 (出題 0 問) は record/streak が入力と同値なので保存を省く。
        if (!update.didRecord) return update
        _records.value = _records.value + (kind to update.record)
        saveRecords(_records.value)
        _streak.value = update.streak
        saveStreak(update.streak)
        return update
    }

    // MARK: - 永続化

    private fun loadRecords(): Map<GameKind, GameRecord> {
        val raw = prefs.getString(KEY_RECORDS, null) ?: return emptyMap()
        return try {
            val json = JSONObject(raw)
            GameKind.entries.mapNotNull { kind ->
                if (!json.has(kind.name)) return@mapNotNull null
                val o = json.getJSONObject(kind.name)
                kind to GameRecord(
                    lastScore = o.optInt("lastScore"),
                    lastOutOf = o.optInt("lastOutOf"),
                    bestScore = o.optInt("bestScore"),
                    bestOutOf = o.optInt("bestOutOf"),
                    playCount = o.optInt("playCount")
                )
            }.toMap()
        } catch (e: Exception) {
            emptyMap()
        }
    }

    private fun saveRecords(records: Map<GameKind, GameRecord>) {
        val json = JSONObject()
        records.forEach { (kind, rec) ->
            json.put(kind.name, JSONObject().apply {
                put("lastScore", rec.lastScore)
                put("lastOutOf", rec.lastOutOf)
                put("bestScore", rec.bestScore)
                put("bestOutOf", rec.bestOutOf)
                put("playCount", rec.playCount)
            })
        }
        prefs.edit().putString(KEY_RECORDS, json.toString()).apply()
    }

    private fun loadStreak(): GameStreakState {
        val raw = prefs.getString(KEY_STREAK, null) ?: return emptyStreakState()
        return try {
            val o = JSONObject(raw)
            GameStreakState(
                streak = o.optInt("streak"),
                totalDays = o.optInt("totalDays"),
                lastClearedDay = o.optString("lastClearedDay").ifEmpty { null }
            )
        } catch (e: Exception) {
            emptyStreakState()
        }
    }

    private fun saveStreak(s: GameStreakState) {
        val json = JSONObject().apply {
            put("streak", s.streak)
            put("totalDays", s.totalDays)
            put("lastClearedDay", s.lastClearedDay ?: "")
        }
        prefs.edit().putString(KEY_STREAK, json.toString()).apply()
    }

    companion object {
        private const val PREFS_NAME = "game_progress_store"
        private const val KEY_RECORDS = "game_records_v1"
        private const val KEY_STREAK = "game_streak_v1"
    }
}

File: ImasLiveDB/Views/Games/GameProgressStore.swift
import Foundation

// =============================================================================
// ゲームの軽量プログレス永続化 (UserDefaults)。サーバ非依存・端末ローカルのみ。
// - 各ゲームの「直近スコア」「最高スコア」「プレイ回数」を記録 → ハブのカードに表示。
// - 「デイリーチャレンジ」= 1日1回どれかのゲームを遊んだら達成。連続達成日数 (ストリーク) を数える。
//   月曜ミーム通知と同じく「毎日開く理由」を作るのが狙い。
//
// 更新規則 (自己ベストの比較・連続達成の加算・記録しない条件) は imas-core の
// `domain/game_progress.rs` が単独で持つ。ここに残すのは **保存の実体** だけで、
// ストアは「読む → コアに渡す → 返ってきた値を書く」に徹する
// (同じ規則を iOS/Android の 2 実装に置くと片方だけ直った時に食い違うため)。
// =============================================================================

/// ハブが束ねるゲームの識別子。rawValue は永続キー兼用なので変更しない。
enum GameKind: String, CaseIterable, Codable, Sendable {
    case introDon
    case idolQuiz
    case songSingerQuiz
    case colorMatch

    /// 表示名 (リザルト・シェア文言で使う)。
    var displayName: String {
        switch self {
        case .introDon:       return "イントロドン"
        case .idolQuiz:       return "アイドル当てクイズ"
        case .songSingerQuiz: return "ソロ曲クイズ"
        case .colorMatch:     return "カラーマッチ"
        }
    }

    /// 0–100 の正規化スコア (正答率) を扱うゲームか。それ以外は「点」をそのまま表示。
    var scoreIsPercent: Bool {
        switch self {
        case .colorMatch: return true
        case .introDon, .idolQuiz, .songSingerQuiz: return false
        }
    }
}

// 1 ゲーム分の記録 `GameRecord` と連続記録 `GameStreakState` の型は imas-core の
// 生成バインディングにある (保存キー game_records_v1 / game_streak_v1 の中身そのもの)。

extension GameRecord {
    /// 1 度でも遊んだか (ハブのカードが「未プレイ」を出すかの判定)。
    /// コアの `has_played()` と同義だが、保存値 1 つを見るだけの述語に FFI 往復を
    /// 足す価値がないので FFI 面には出ていない。表示分岐用にここで持つ。
    var hasPlayed: Bool { playCount > 0 }
}

/// ゲーム横断のローカル進捗ストア。
@Observable @MainActor
final class GameProgressStore {
    static let shared = GameProgressStore()

    private let recordsKey = "game_records_v1"
    private let streakKey = "game_streak_v1"

    /// ゲーム別レコード。
    private(set) var records: [GameKind: GameRecord] = [:]

    /// 連続デイリー達成の保存値 (連続日数・通算日数・最後に達成した端末ローカル日)。
    private(set) var streakState = GameStreakState(streak: 0, totalDays: 0, lastClearedDay: nil)

    private init() { load() }

    // MARK: - 参照

    /// 未プレイのゲームの初期値 (コアの `GameRecord::default()` と同じゼロ値)。
    private static let emptyRecord = GameRecord(lastScore: 0, lastOutOf: 0, bestScore: 0,
                                                bestOutOf: 0, playCount: 0)

    func record(for kind: GameKind) -> GameRecord { records[kind] ?? Self.emptyRecord }

    /// 自己ベストの正答率 (0–100)。まだ記録が無ければ nil
    /// (「—」を出すか今回の率で代用するかは画面ごとに違うので、文言には落とさない)。
    func bestRatePercent(for kind: GameKind) -> Int? {
        gameProgressBestRatePercent(record: record(for: kind)).map(Int.init)
    }

    /// 連続デイリー達成日数 (保存値そのもの)。表示には `displayStreak` を使う。
    var streak: Int { Int(streakState.streak) }
    /// 通算デイリー達成日数。
    var totalDays: Int { Int(streakState.totalDays) }
    /// 最後にデイリーを達成した日 (端末ローカル YYYY-MM-DD)。未達成は nil。
    var lastClearedDay: String? { streakState.lastClearedDay }

    /// 今日デイリーチャレンジを達成済みか。
    var didClearToday: Bool {
        gameProgressDidClearToday(streak: streakState, todayKey: DailyPick.dayKey())
    }

    /// ストリークが「今日途切れていないか」を織り込んだ表示用の連続日数。
    var displayStreak: Int {
        Int(gameProgressDisplayStreak(streak: streakState,
                                      todayKey: DailyPick.dayKey(),
                                      yesterdayKey: DailyPick.previousDayKey()))
    }

    // MARK: - 記録

    /// ゲーム結果を記録する。score/outOf は「正解ポイント / 満点」。
    /// 同時に当日のデイリーチャレンジ達成 + ストリーク更新を行う。
    ///
    /// 戻り値の `isNewBest` が結果画面の「自己ベスト更新！」バッジ。判定は best を
    /// 上書きする前に済ませる必要があり、その順序ごとコア側で固定されている
    /// (画面が「読む → 判定 → 書く」を自前で並べると順序を崩した瞬間に恒久的に false になる)。
    @discardableResult
    func recordResult(_ kind: GameKind, score: Int, outOf: Int) -> GameProgressUpdate {
        // 連続クリア日数は「そのユーザーの 1 日」が単位なので端末ローカル日 (`DailyPick`)。
        // 公演日との比較に使う `JSTDay` (JST 固定) とは意味が違う。
        let update = gameProgressApplyResult(
            record: record(for: kind),
            streak: streakState,
            score: Int32(clamping: score),
            outOf: Int32(clamping: outOf),
            todayKey: DailyPick.dayKey(),
            yesterdayKey: DailyPick.previousDayKey())
        // 記録として成立しなかった場合 (出題 0 問) は保存値が入力と同値なので書かない。
        guard update.didRecord else { return update }
        records[kind] = update.record
        streakState = update.streak
        save()
        return update
    }

    // MARK: - 永続化

    /// UserDefaults に入っている JSON の形。コアの `GameRecord` は FFI 型で Codable を
    /// 持たないので、出荷済みの保存形 (キー名・順序) を固定する DTO をここに置く。
    /// フィールド名を変えると既存ユーザの記録が読めなくなる。
    private struct StoredRecord: Codable {
        var lastScore: Int
        var lastOutOf: Int
        var bestScore: Int
        var bestOutOf: Int
        var playCount: Int

        init(_ r: GameRecord) {
            lastScore = Int(r.lastScore)
            lastOutOf = Int(r.lastOutOf)
            bestScore = Int(r.bestScore)
            bestOutOf = Int(r.bestOutOf)
            playCount = Int(r.playCount)
        }

        var record: GameRecord {
            GameRecord(lastScore: Int32(clamping: lastScore), lastOutOf: Int32(clamping: lastOutOf),
                       bestScore: Int32(clamping: bestScore), bestOutOf: Int32(clamping: bestOutOf),
                       playCount: Int32(clamping: playCount))
        }
    }

    /// 保存キー game_streak_v1 の中身 (`GameStreakState` の保存形)。
    private struct StreakState: Codable {
        var streak: Int
        var totalDays: Int
        var lastClearedDay: String?

        init(_ s: GameStreakState) {
            streak = Int(s.streak)
            totalDays = Int(s.totalDays)
            lastClearedDay = s.lastClearedDay
        }

        var state: GameStreakState {
            GameStreakState(streak: Int32(clamping: streak), totalDays: Int32(clamping: totalDays),
                            lastClearedDay: lastClearedDay)
        }
    }

    private func load() {
        if let data = UserDefaults.standard.data(forKey: recordsKey),
           let decoded = try? JSONDecoder().decode([GameKind: StoredRecord].self, from: data) {
            records = decoded.mapValues(\.record)
        }
        if let data = UserDefaults.standard.data(forKey: streakKey),
           let s = try? JSONDecoder().decode(StreakState.self, from: data) {
            streakState = s.state
        }
    }

    private func save() {
        if let data = try? JSONEncoder().encode(records.mapValues(StoredRecord.init)) {
            UserDefaults.standard.set(data, forKey: recordsKey)
        }
        if let data = try? JSONEncoder().encode(StreakState(streakState)) {
            UserDefaults.standard.set(data, forKey: streakKey)
        }
    }
}
